### ISSUE TITLE
Emulate INSERT .. ON DUPLICATE KEY UPDATE .. WHERE on MySQL

### ISSUE THREAD

**Issue by kjeet** [jOOQ/jOOQ#8479]
Using jooq 3.11.9, the following query prints:
```
    public static void main(String[] args) throws SQLException {
        String sql = DSL.using(SQLDialect.MYSQL_5_7).renderInlined(
                insertInto(table("tableA"))
                        .columns(field("colA"), field("colB"))
                        .values("val1", "val2")
                        .onDuplicateKeyUpdate()
                        .set(field("colA"), "something")
                        .where(field("colC").eq("other")));
        System.out.println(sql);
    }
```
`insert into tableA (colA, colB) values ('val1', 'val2') on duplicate key update colA = 'something'`

The WHERE clause and "colC" are never mentioned in the query. I'm might be inexperienced with SQL but I would expect and exception or at least a warning would be thrown if I'm doing something invalid or unsupported.

**Comment by lukaseder** [comment]
Thank you very much for your request. MySQL does not support a `WHERE` clause in the `ON DUPLICATE KEY UPDATE` syntax. Please not that there's no `SQLDialect.MYSQL` reference on the `InsertOnConflictWhereStep.where()` methods.

But we could emulate it. The SQL that should be generated is:

```sql
INSERT INTO tableA (colA, colB)
VALUES ('val1', 'val2')
ON DUPLICATE KEY UPDATE colA = CASE WHEN colC = 'other' THEN 'something' ELSE colA
```

The result would be the same, although the update count returned from this statement might be different from e.g. PostgreSQL.

**Comment by kjeet** [comment]
Thanks! For future reference I've ended up doing the following which becomes the `CASE WHEN` clause. (Which is a little weird since MySQL support IF() but it works!)

```
        String sql = DSL.using(SQLDialect.MYSQL_5_7).renderInlined(
                insertInto(table("tableA"))
                        .columns(field("colA"), field("colB"))
                        .values("val1", "val2")
                        .onDuplicateKeyUpdate()
                        .set(field("colA"),
                                iif(field("colC").eq("other"), "something", field("colA", String.class))
                            ));
        System.out.println(sql);
```
Thanks!

**Comment by lukaseder** [comment]
@kjeet Indeed, you can obviously implement my suggested emulation manually.

Note `iif()` is the SQL Server style function: https://docs.microsoft.com/en-us/sql/t-sql/functions/logical-functions-iif-transact-sql?view=sql-server-2017

We're not supporting MySQL's `IF()` function yet in the API. We might, in the future.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.12
